### DISCUSSION TITLE
A simple stress test of apisix and nginx

### DISCUSSION THREAD

**Discussion by Lewisyixin** [apache/apisix#7204] [Q&A]
I did a simple stress test of apisix&nginx, and found performance of apisix is little better than nginx, is it normal?
In my opinion, performance of  apisix should be a little  worse than nginx.

nginx and apisix are on the same vm and proxy_pass to the same upstream
nginx:
version: 1.20.1
config:
![image](https://user-images.githubusercontent.com/17898688/172316225-47461afd-7d09-425f-8bf1-b691c2073066.png)
result:
![image](https://user-images.githubusercontent.com/17898688/172317039-88cd1be0-8725-43bf-a79e-8a6eb8aa42ec.png)

apisix
version: 2.14.1
![image](https://user-images.githubusercontent.com/17898688/172316708-f2abfeab-867f-4853-a3b8-17e7708eac6a.png)
result:
![image](https://user-images.githubusercontent.com/17898688/172316979-6e96cf82-893c-48fd-bb65-febdcd41259e.png)

**Comment by moonming** [comment]
During the performance test, are the CPUs of nginx and Apache APISIX fully loaded?

**Reply by tokers** [reply]
Also make sure the machines for running APISIX and Nginx are same and the number of worker processes should be same.

**Reply by Lewisyixin** [reply]
I use "worker_processes auto;"  on the last performance test(both nginx and apisix and they are on the same machine). 
I did a test again today. First I change there "worker_processes auto;" to "worker_processes 1;", and got samilar result.
Then I increased concurrency to make sure cpu would be fully loaed as much as possible. 
The result is that performance of apisix  is still a little worse than nginx but load of nginx is lower than apisix.
Is this as expected？


nginx
![image](https://user-images.githubusercontent.com/17898688/173039816-047e4f19-8b78-473c-b47a-c548bc04c075.png)
![image](https://user-images.githubusercontent.com/17898688/173040871-690dd963-ae0b-45b6-84f0-92a1968c58cb.png)
![image](https://user-images.githubusercontent.com/17898688/173040927-b4cb5e92-1d86-4eae-9e2c-bc3a80db868a.png)


apisix
![image](https://user-images.githubusercontent.com/17898688/173040620-2e9d5add-8db9-4f65-8d16-7dc3446529e4.png)
![image](https://user-images.githubusercontent.com/17898688/173040547-f5645e22-1041-45f6-a479-af7558a3a32b.png)
![image](https://user-images.githubusercontent.com/17898688/173040588-f9192238-24b5-484c-a9ee-cc193a1515fd.png)

**Reply by tokers** [reply]
Looks like you don't enable keep-alive for Nginx. Add following two directives in your Nginx conf.

```
proxy_http_version 1.1;
proxy_set_header Connection "";
```

**Reply by Lewisyixin** [reply]
Oh, that's right!  Performance of nginx is a bit higher than apisix after enable keep-alive.
